Add common scalar reads to OrcInputStream

oerling opened the pull request:
Adds an optimized read path for varints, floats and doubles. This
avoids copying and bounds checking for consecutive bytes.

Removes the use of Slice from OrcInputStream. Keeps the decompression
buffer separate from the Slice read from the compressed input. The
backing byte[] of the compressed input Slice is directly used as the
buffer when there is no compression, avoiding needless copy.

```
== RELEASE NOTES ==

Hive Changes
--------------
 * Streamline ORC read path for varints and fixed width numbers.
```

oerling commented:
Thank you for the review.

I updated the PR. Do you mean we should split the commits now?

 

 

Release note?

 

A: Any example of usage for this?

 

 

 

I don't understand why you call sliceInput.setPosition(sliceInputPosition); - Would you explain?

 

 

A: This was the original from Elon. I take it the intent was not to have a side effect on the input. Another behavior would be to leave the slice input at end.

 

Neither exactly corresponds to the original, which delays the side effect on the input to the time of reading. This too could be reproduced if we wanted to.

 

Considering the use case, OrcInputStream takes ownership of the input and it can do the side effect at any or no time. So, leaving the stream at end is just as good and less code. There are no cases where any other reader from the slice input could/should coexist with the OrcInputStream, so on balance of safety leaving the stream at end could be seen as best.

 

 

 

 

 

 

 

 

 

 

 

From: Maria Basmanova <notifications@github.com> 
Sent: Monday, September 23, 2019 9:09 AM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Add common scalar reads to OrcInputStream (#13434)

 

@mbasmanova requested changes on this pull request.

@oerling <https://github.com/oerling>  Fantastic! This is such a powerful optimization and it applies to both default and Aria scans. I ran benchmarks for both scans for floats and doubles and I'm seeing 2-3x improvement. Thank you!

(Looks like our benchmarks don't cover DWRF longs. We should add these.)

---- with optimization ----
 
# Run complete. Total time: 00:03:25
 
 
 
Benchmark                             (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
 
BenchmarkSelectiveStreamReaders.read             real         true  avgt   20  0.047 ± 0.001   s/op
 
BenchmarkSelectiveStreamReaders.read             real        false  avgt   20  0.041 ± 0.002   s/op
 
BenchmarkSelectiveStreamReaders.read           double         true  avgt   20  0.061 ± 0.003   s/op
 
BenchmarkSelectiveStreamReaders.read           double        false  avgt   20  0.064 ± 0.004   s/op
 
 
 
 
 
 
 
# Run complete. Total time: 00:05:10
 
 
 
Benchmark                               (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
 
BenchmarkBatchStreamReaders.readBlocks             real      PARTIAL  avgt   30  0.049 ± 0.002   s/op
 
BenchmarkBatchStreamReaders.readBlocks             real         NONE  avgt   30  0.048 ± 0.001   s/op
 
BenchmarkBatchStreamReaders.readBlocks           double      PARTIAL  avgt   30  0.062 ± 0.002   s/op
 
BenchmarkBatchStreamReaders.readBlocks           double         NONE  avgt   30  0.070 ± 0.002   s/op
 
 
 
 
 
---- without ----
 
# Run complete. Total time: 00:03:34
 
 
 
Benchmark                             (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
 
BenchmarkSelectiveStreamReaders.read             real         true  avgt   20  0.123 ± 0.003   s/op
 
BenchmarkSelectiveStreamReaders.read             real        false  avgt   20  0.202 ± 0.005   s/op
 
BenchmarkSelectiveStreamReaders.read           double         true  avgt   20  0.126 ± 0.004   s/op
 
BenchmarkSelectiveStreamReaders.read           double        false  avgt   20  0.218 ± 0.013   s/op
 
 
 
# Run complete. Total time: 00:03:34
 
 
 
Benchmark                               (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
 
BenchmarkBatchStreamReaders.readBlocks             real      PARTIAL  avgt   20  0.127 ± 0.003   s/op
 
BenchmarkBatchStreamReaders.readBlocks             real         NONE  avgt   20  0.200 ± 0.007   s/op
 
BenchmarkBatchStreamReaders.readBlocks           double      PARTIAL  avgt   20  0.133 ± 0.006   s/op
 
BenchmarkBatchStreamReaders.readBlocks           double         NONE  avgt   20  0.203 ± 0.006   s/op
 

Next time, let's break up such changes into multiple commits. Here, I'd do the following:

*	commit with OrcInputStream optimizations to remove current Slice.
*	commit with the new readDouble method and corresponding changes in DoubleInputStream
*	commit with the new readFloat method and corresponding changes in FloadInputStream
*	all the changes to optimize reading longs

This change affects default scan path, hence, we need to (1) add release nodes to PR description; (2) run verifier for different types of production workload to make sure nothing breaks.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java <https://github.com/prestodb/presto/pull/13434#discussion_r327177962> :

> @@ -71,12 +89,16 @@ public OrcInputStream(
 
         systemMemoryContext.newLocalMemoryContext(OrcInputStream.class.getSimpleName()).setBytes(sliceInputRetainedSizeInBytes);
 
 
 
         if (!decompressor.isPresent()) {
 
-            this.current = sliceInput;
 
+            long sliceInputPosition = sliceInput.position();
 

*	sliceInput.position() and sliceInput.remaining() is called multiple times in this block; I suggest to re-write it to use these only once
*	I don't understand why you call sliceInput.setPosition(sliceInputPosition); - Would you explain?

            this.length = toIntExact(sliceInput.remaining());
 
            this.buffer = new byte[this.length];
 
 
 
            int sliceInputPosition = toIntExact(sliceInput.position());
 
            sliceInput.readFully(buffer, sliceInputPosition, this.length);
 
            sliceInput.setPosition(sliceInputPosition);
 
            
 
            this.compressedSliceInput = EMPTY_SLICE.getInput();
 
  _____  


In presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java <https://github.com/prestodb/presto/pull/13434#discussion_r327184219> :

> +    public void skipVarints(long items)
 
+            throws IOException
 
+    {
 
+        if (items == 0) {
 
+            return;
 
+        }
 
+
 
+        while (items > 0) {
 
+            items -= skipVarintsInBuffer(items);
 
+        }
 
+    }
 
+
 
+    private long skipVarintsInBuffer(long items)
 
+            throws IOException
 
+    {
 
+        if (items == 0) {
 

this check is always false; let's remove it

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java <https://github.com/prestodb/presto/pull/13434#discussion_r327185132> :

> +    }
 
+
 
+    public float readFloat()
 
+            throws IOException
 
+    {
 
+        int readPosition = ensureContiguousBytesAndAdvance(SIZE_OF_FLOAT);
 
+        if (readPosition < 0) {
 
+            return ByteArrays.getFloat(tempDouble, 0);
 
+        }
 
+        return ByteArrays.getFloat(buffer, readPosition);
 
+    }
 
+
 
+    private int ensureContiguousBytesAndAdvance(int numBytes)
 
+            throws IOException
 
+    {
 
+        // If there are numBytes in the buffer, return the offsett of the start and advances by numBytes. If not, copies numBytes into tempDouble and advamces numBytes and returns -1.
 

typos: offsett -> offset, advances -> advance; copies -> copy; advamces -> advance by; etc.

// If there are numBytes in the buffer, return the offset of the start and advance by numBytes. If not, copy numBytes into tempDouble, advance by numBytes and return -1.
 

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13434?email_source=notifications&email_token=AKPPPTYIIQUVS5ET6XILWVTQLDSZHA5CNFSM4IZGIQ72YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFSW6UA#pullrequestreview-291860304> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPT5SVV2SVEWVXNSTCOTQLDSZHANCNFSM4IZGIQ7Q> .

mbasmanova commented:
@oerling 

> Do you mean we should split the commits now?

Not unless this is relatively easy. I was primarily commented for the future PRs.

Re: sliceInput.setPosition - it was also my understanding that this is unnecessary. Thanks for confirming and remove that code.

Re: release notes - Here is an example. I need a one-liner that describes this change for the users.

oerling commented:
Hi Masha

 

Splitting the commit will just add errors. Next time.

 

Ying added a line to the release notes.

 

I started a verifier run with only this change in now. We’ll get some 5K queries done before midnight kills it. This same code has been in the prototype runs for all time so it works, unless it was broken in transfer, which I doubt. We’ll see.

 

 

 

From: Maria Basmanova <notifications@github.com> 
Sent: Monday, September 23, 2019 1:36 PM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Add common scalar reads to OrcInputStream (#13434)

 

@oerling <https://github.com/oerling> 

Do you mean we should split the commits now?

Not unless this is relatively easy. I was primarily commented for the future PRs.

Re: sliceInput.setPosition - it was also my understanding that this is unnecessary. Thanks for confirming and remove that code.

Re: release notes - Here is an example. I need a one-liner that describes this change for the users.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13434?email_source=notifications&email_token=AKPPPT62YKS6G2YBHVXCQYDQLESDDA5CNFSM4IZGIQ72YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7MFZMY#issuecomment-534273203> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPT3Z7CNE3B6BB3PXFITQLESDDANCNFSM4IZGIQ7Q> .

oerling commented:
Let’s hold this for a bit while we fix a bug. We have easy reproduction based on a sample of data from prod. 

 

 

From: Maria Basmanova <notifications@github.com> 
Sent: Monday, September 23, 2019 1:37 PM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Add common scalar reads to OrcInputStream (#13434)

 

@mbasmanova approved this pull request.

@oerling <https://github.com/oerling>  Looks good to me. Thanks for identifying and implementing this optimization.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13434?email_source=notifications&email_token=AKPPPTZC6WIPMTOGJI5S7FDQLESFHA5CNFSM4IZGIQ72YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFUCKPI#pullrequestreview-292037949> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPT7BTJ5EPS5RJWOFBP3QLESFHANCNFSM4IZGIQ7Q> .

oerling commented:
OK, it appears fixed, first 5% of the run come out with no errors, as was not the case before. The error is a new case that was not in the prototype. Will add a unit test to cover the mix of values that make this and update tomorrow.

 

 

From: Maria Basmanova <notifications@github.com> 
Sent: Monday, September 23, 2019 1:37 PM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Add common scalar reads to OrcInputStream (#13434)

 

@mbasmanova approved this pull request.

@oerling <https://github.com/oerling>  Looks good to me. Thanks for identifying and implementing this optimization.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13434?email_source=notifications&email_token=AKPPPTZC6WIPMTOGJI5S7FDQLESFHA5CNFSM4IZGIQ72YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFUCKPI#pullrequestreview-292037949> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPT7BTJ5EPS5RJWOFBP3QLESFHANCNFSM4IZGIQ7Q> .

mbasmanova commented on presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java:
```diff
@@ -1007,4 +1016,15 @@ private static String toCharValue(Object value, int minLength)
+    {
+        List<Long> values = new ArrayList<>();
+        long[] numbers = new long[] {1L, 1L << 8, 1L << 13, 1L << 20, 1L << 27, 1L << 34, 1L << 40, 1L << 47, 1L << 53, 1L << 60, 1L << 63};
+        for (int i = 0; i < 30000; i++) {
```
instead of hard-coding number of rows, add an `int numRows` parameter and call it with `NUM_ROWS`

```
        List<Long> values = varintScaleSequence(NUM_ROWS);
...
    private static List<Long> varintScaleSequence(int numRows)
    {
        List<Long> values = new ArrayList<>();
        long[] numbers = new long[] {1L, 1L << 8, 1L << 13, 1L << 20, 1L << 27, 1L << 34, 1L << 40, 1L << 47, 1L << 53, 1L << 60, 1L << 63};
        for (int i = 0; i < numRows; i++) {
            values.add(numbers[i % numbers.length] + i);
            values.add(-numbers[i % numbers.length] + i);
        }
        return values;
    }
```

oerling commented:
Line 332 in OrcInputStream. The first loop ends after 1 + 7 bytes. One must distinguish between the ending because of exhausting the word vs ending because the last byte was hit. So varints of exactly 8 bytes would work incorrectly. 

 

 

 

From: Maria Basmanova <notifications@github.com> 
Sent: Friday, September 27, 2019 1:20 AM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Add common scalar reads to OrcInputStream (#13434)

 

@mbasmanova commented on this pull request.

@oerling <https://github.com/oerling>  Orri, can you explain what was the bug and what was the fix? Which lines changed?

  _____  

In presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java <https://github.com/prestodb/presto/pull/13434#discussion_r328960530> :

> @@ -1007,4 +1016,15 @@ private static String toCharValue(Object value, int minLength)
         }
         return values;
     }
+
+    private static List<Long> varintScaleSequence()
+    {
+        List<Long> values = new ArrayList<>();
+        long[] numbers = new long[] {1L, 1L << 8, 1L << 13, 1L << 20, 1L << 27, 1L << 34, 1L << 40, 1L << 47, 1L << 53, 1L << 60, 1L << 63};
+        for (int i = 0; i < 30000; i++) {

instead of hard-coding number of rows, add an int numRows parameter and call it with NUM_ROWS

        List<Long> values = varintScaleSequence(NUM_ROWS);
...
    private static List<Long> varintScaleSequence(int numRows)
    {
        List<Long> values = new ArrayList<>();
        long[] numbers = new long[] {1L, 1L << 8, 1L << 13, 1L << 20, 1L << 27, 1L << 34, 1L << 40, 1L << 47, 1L << 53, 1L << 60, 1L << 63};
        for (int i = 0; i < numRows; i++) {
            values.add(numbers[i % numbers.length] + i);
            values.add(-numbers[i % numbers.length] + i);
        }
        return values;
    }

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13434?email_source=notifications&email_token=AKPPPT45FRQHBPDA7UCJIW3QLW63TA5CNFSM4IZGIQ72YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCGEMDFI#pullrequestreview-294175125> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPT2UH6QT2BSOVWQKHE3QLW63TANCNFSM4IZGIQ7Q> .

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java:
```diff
@@ -45,9 +44,14 @@
+    private byte[] decompressionResultBuffer;
+    private int position;
+    private int length;
+    private boolean isUncompressed;
```
This can be a local variable. No need to put on stack.

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java:
```diff
@@ -241,6 +247,40 @@ public long skip(long n)
+        return ByteArrays.getFloat(buffer, readPosition);
+    }
+
+    private int ensureContiguousBytesAndAdvance(int numBytes)
```
nit: numBytes -> bytes

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java:
```diff
@@ -52,6 +55,9 @@
     private int uncompressedOffset;
 
+    // Temporary memory for reading a float or double at buffer boundary.
+    private byte[] tempDouble = new byte[SIZE_OF_DOUBLE];
```
nit: let's call this temporaryBuffer.

highker commented on presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java:
```diff
@@ -1007,4 +1016,15 @@ private static String toCharValue(Object value, int minLength)
         return values;
     }
+
+    private static List<Long> varintScaleSequence(int numRows)
```
nit: numRows -> rows

highker commented on presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java:
```diff
@@ -1007,4 +1016,15 @@ private static String toCharValue(Object value, int minLength)
+
+    private static List<Long> varintScaleSequence(int numRows)
+    {
+        List<Long> values = new ArrayList<>();
```
Let's use `ImmutableList.builder`

oerling commented:
I addressed the comments. 

The List<Long> in the test must be mutable. All the others are that too. They get shuffled later on.

Thanks

Orri

 

 

 

From: James Sun <notifications@github.com> 
Sent: Thursday, October 3, 2019 2:34 PM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Add common scalar reads to OrcInputStream (#13434)

 

@highker commented on this pull request.

There is a typo in the second commit "flots" -> "floats". Otherwise, looks good to me.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java <https://github.com/prestodb/presto/pull/13434#discussion_r331257514> :

>  
     private byte[] buffer;
+    private byte[] decompressionResultBuffer;
+    private int position;
+    private int length;
+    private boolean isUncompressed;

This can be a local variable. No need to put on stack.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java <https://github.com/prestodb/presto/pull/13434#discussion_r331258728> :

> +            return ByteArrays.getDouble(tempDouble, 0);
+        }
+        return ByteArrays.getDouble(buffer, readPosition);
+    }
+
+    public float readFloat()
+            throws IOException
+    {
+        int readPosition = ensureContiguousBytesAndAdvance(SIZE_OF_FLOAT);
+        if (readPosition < 0) {
+            return ByteArrays.getFloat(tempDouble, 0);
+        }
+        return ByteArrays.getFloat(buffer, readPosition);
+    }
+
+    private int ensureContiguousBytesAndAdvance(int numBytes)

nit: numBytes -> bytes

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java <https://github.com/prestodb/presto/pull/13434#discussion_r331259333> :

> @@ -52,6 +55,9 @@
     private boolean isUncompressed;
     private int uncompressedOffset;
 
+    // Temporary memory for reading a float or double at buffer boundary.
+    private byte[] tempDouble = new byte[SIZE_OF_DOUBLE];

nit: let's call this temporaryBuffer.

  _____  

In presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java <https://github.com/prestodb/presto/pull/13434#discussion_r331261201> :

> @@ -1007,4 +1016,15 @@ private static String toCharValue(Object value, int minLength)
         }
         return values;
     }
+
+    private static List<Long> varintScaleSequence(int numRows)

nit: numRows -> rows

  _____  

In presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java <https://github.com/prestodb/presto/pull/13434#discussion_r331262863> :

> @@ -1007,4 +1016,15 @@ private static String toCharValue(Object value, int minLength)
         }
         return values;
     }
+
+    private static List<Long> varintScaleSequence(int numRows)
+    {
+        List<Long> values = new ArrayList<>();

Let's use ImmutableList.builder

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13434?email_source=notifications&email_token=AKPPPT7B77FARVVJFXOEMLDQMZQK5A5CNFSM4IZGIQ72YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCG3EDRA#pullrequestreview-297157060> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPTYVRAZILJGQDTVYKV3QMZQK5ANCNFSM4IZGIQ7Q> .

oerling commented:
There  may be some weird bug that occurs with this so do not merge yet.

 

 

 

 

From: James Sun <notifications@github.com> 
Sent: Thursday, October 3, 2019 2:34 PM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Add common scalar reads to OrcInputStream (#13434)

 

@highker commented on this pull request.

There is a typo in the second commit "flots" -> "floats". Otherwise, looks good to me.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java <https://github.com/prestodb/presto/pull/13434#discussion_r331257514> :

>  
     private byte[] buffer;
+    private byte[] decompressionResultBuffer;
+    private int position;
+    private int length;
+    private boolean isUncompressed;

This can be a local variable. No need to put on stack.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java <https://github.com/prestodb/presto/pull/13434#discussion_r331258728> :

> +            return ByteArrays.getDouble(tempDouble, 0);
+        }
+        return ByteArrays.getDouble(buffer, readPosition);
+    }
+
+    public float readFloat()
+            throws IOException
+    {
+        int readPosition = ensureContiguousBytesAndAdvance(SIZE_OF_FLOAT);
+        if (readPosition < 0) {
+            return ByteArrays.getFloat(tempDouble, 0);
+        }
+        return ByteArrays.getFloat(buffer, readPosition);
+    }
+
+    private int ensureContiguousBytesAndAdvance(int numBytes)

nit: numBytes -> bytes

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java <https://github.com/prestodb/presto/pull/13434#discussion_r331259333> :

> @@ -52,6 +55,9 @@
     private boolean isUncompressed;
     private int uncompressedOffset;
 
+    // Temporary memory for reading a float or double at buffer boundary.
+    private byte[] tempDouble = new byte[SIZE_OF_DOUBLE];

nit: let's call this temporaryBuffer.

  _____  

In presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java <https://github.com/prestodb/presto/pull/13434#discussion_r331261201> :

> @@ -1007,4 +1016,15 @@ private static String toCharValue(Object value, int minLength)
         }
         return values;
     }
+
+    private static List<Long> varintScaleSequence(int numRows)

nit: numRows -> rows

  _____  

In presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java <https://github.com/prestodb/presto/pull/13434#discussion_r331262863> :

> @@ -1007,4 +1016,15 @@ private static String toCharValue(Object value, int minLength)
         }
         return values;
     }
+
+    private static List<Long> varintScaleSequence(int numRows)
+    {
+        List<Long> values = new ArrayList<>();

Let's use ImmutableList.builder

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13434?email_source=notifications&email_token=AKPPPT7B77FARVVJFXOEMLDQMZQK5A5CNFSM4IZGIQ72YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCG3EDRA#pullrequestreview-297157060> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPTYVRAZILJGQDTVYKV3QMZQK5ANCNFSM4IZGIQ7Q> .

oerling commented:
Here is the verifier testing record for this PR.

 

 

Run 22246 of ftw2  - column mismatches nondeterministic, top k, quantiles

Run 22251 - column mismatches , top k

Run 22328 on vll1_verifier1  - 2 nondeterministic top k, 1 row mismatch, rechecked, not reproduced after sync to latest master.

Run 22330 on ftw3_verifier1 - 6 nondeterministic top k, 2 WS failures

 

22498 vll1_verifier1 1 ws file missing, 5 nondeterministic top k, 1 sum of double variability from aggregation order. 

 

I have checked the suspect cases and the nondeterminism is explained as above. As far as I am concerned you can merge this.

 

 

 

 

 

From: James Sun <notifications@github.com> 
Sent: Thursday, October 3, 2019 2:34 PM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Add common scalar reads to OrcInputStream (#13434)

 

@highker commented on this pull request.

There is a typo in the second commit "flots" -> "floats". Otherwise, looks good to me.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java <https://github.com/prestodb/presto/pull/13434#discussion_r331257514> :

>  
     private byte[] buffer;
+    private byte[] decompressionResultBuffer;
+    private int position;
+    private int length;
+    private boolean isUncompressed;

This can be a local variable. No need to put on stack.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java <https://github.com/prestodb/presto/pull/13434#discussion_r331258728> :

> +            return ByteArrays.getDouble(tempDouble, 0);
+        }
+        return ByteArrays.getDouble(buffer, readPosition);
+    }
+
+    public float readFloat()
+            throws IOException
+    {
+        int readPosition = ensureContiguousBytesAndAdvance(SIZE_OF_FLOAT);
+        if (readPosition < 0) {
+            return ByteArrays.getFloat(tempDouble, 0);
+        }
+        return ByteArrays.getFloat(buffer, readPosition);
+    }
+
+    private int ensureContiguousBytesAndAdvance(int numBytes)

nit: numBytes -> bytes

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java <https://github.com/prestodb/presto/pull/13434#discussion_r331259333> :

> @@ -52,6 +55,9 @@
     private boolean isUncompressed;
     private int uncompressedOffset;
 
+    // Temporary memory for reading a float or double at buffer boundary.
+    private byte[] tempDouble = new byte[SIZE_OF_DOUBLE];

nit: let's call this temporaryBuffer.

  _____  

In presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java <https://github.com/prestodb/presto/pull/13434#discussion_r331261201> :

> @@ -1007,4 +1016,15 @@ private static String toCharValue(Object value, int minLength)
         }
         return values;
     }
+
+    private static List<Long> varintScaleSequence(int numRows)

nit: numRows -> rows

  _____  

In presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java <https://github.com/prestodb/presto/pull/13434#discussion_r331262863> :

> @@ -1007,4 +1016,15 @@ private static String toCharValue(Object value, int minLength)
         }
         return values;
     }
+
+    private static List<Long> varintScaleSequence(int numRows)
+    {
+        List<Long> values = new ArrayList<>();

Let's use ImmutableList.builder

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13434?email_source=notifications&email_token=AKPPPT7B77FARVVJFXOEMLDQMZQK5A5CNFSM4IZGIQ72YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCG3EDRA#pullrequestreview-297157060> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPTYVRAZILJGQDTVYKV3QMZQK5ANCNFSM4IZGIQ7Q> .

oerling commented:
Edited the messages.

 

 

 

From: James Sun <notifications@github.com> 
Sent: Tuesday, October 8, 2019 7:30 PM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Add common scalar reads to OrcInputStream (#13434)

 

@highker approved this pull request.

Looks good! @oerling <https://github.com/oerling> , there are extra proceeding indentations in the last two commit message bodies and titles. Could you remove them?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13434?email_source=notifications&email_token=AKPPPT7Y2I2FMZDTAIAJS2TQNU6ZPA5CNFSM4IZGIQ72YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCHKIUUY#pullrequestreview-299141715> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPT3XJBSPINPNPZLCIYTQNU6ZPANCNFSM4IZGIQ7Q> .

bhhari commented:
lgtm
